blog/2024-03-26-intel/index.mdx

```diff
@@ -0,0 +1,185 @@
+
+The runtime complexity for these vector distance operations scales linearly with the dimensionality of the vector. Typically vector dimensionality tends to be anywhere from 256 to 2096 in practice - this number is trending upwards over time - which means that distance calculations are also taking longer to perform.
+
+This being the case, we need to ensure that this operation is done as fast and efficiently as possible, whether through leveraging hardware or software efficiencies.
+
+Let’s take a look at how these calculations can be implemented, for simplicity's sake we will consider the dot product only.
+
+## Vector Distance Implementations
+
+### Naive Go Implementation
+
+The simplest way to implement the dot product in Go is to loop over the vector dimensions one at a time and sum the product of the vectors. This is quite slow for higher dimensionality but it works well for low dimensional vectors.
+
+```Go
+func dotProduct(a, b []float32) float32 {
+
+    var result float32
+
+    for i := 0; i < len(a); i++ {
+
+        result += a[i] * b[i]
+
+    }
+    return result
+}
+```
+
+### Introduction to SIMD
+
+SIMD, which stands for Single Instruction, Multiple Data, is a parallel processing technique that allows a single instruction to operate on multiple data elements simultaneously. This approach is commonly used to improve the performance of various algorithms, including those involving vector and matrix operations.
+
+In the context of vector distance calculations, SIMD instructions can be used to calculate distances between vectors more efficiently. By grouping the dimensions of the vectors into smaller chunks (e.g., 128-bit or 256-bit registers), SIMD instructions can perform the same operation on multiple dimensions in parallel, chunk by chunk. This parallel processing capability can lead to significant performance improvements when dealing with high-dimensional data.
+
+
+![SIMD](./img/SIMD.png)
+
+
+Intel CPUs feature support for advanced SIMD instruction sets, with notable examples being AVX-256 and AVX-512. AVX-256 enables the simultaneous processing of 8 32-bit floating-point numbers (8 dimensions), while AVX-512 doubles this capability to 16 dimensions.
+
+Weaviate uses  Go assembly to generate SIMD instructions. In doing this we give up code readability for performance and so we reserve this technique for distance functions alone where every ounce of performance is needed.
+
+Rather than writing Go assembly directly, we write the code in C and then use the [Go Goat library](https://github.com/gorse-io/goat) which uses Clang/LLVM to generate the final Go assembly. This compiler optimization further improves the performance of the code.
```

Comment:
```suggestion
Rather than writing Go assembly directly, we write the code in C and then use the [Go goat library](https://github.com/gorse-io/goat) which uses Clang/LLVM to generate the final Go assembly. This compiler optimization further improves the performance.
```